quipucords/scanner/job.py

```diff
@@ -147,11 +150,90 @@ class ScanJobRunner:
+
+    CeleryBasedScanJobRunner is superficially similar to SyncScanJobRunner, but there
+    are several important differences to consider. This scan job runner uses Celery to
+    offload processing of scan task work, and it parallelizes that work where possible.
+    """
+
+    def __init__(self, scan_job: ScanJob):
+        """Initialize the job runner."""
+        self.scan_job: ScanJob = scan_job
+
+    def run(self) -> AsyncResult:
+        """Prepare this ScanJob's tasks, and request Celery to run them.
+
+        The net result of this method is *similar to* `SyncScanJobRunner.run`, but this
+        method returns nothing because it delegates most work to Celery tasks that will
+        execute and complete asynchronously.
+
+        Whereas SyncScanJobRunner and ProcessBasedScanJobRunner both work through tasks
+        via a linear queue, CeleryBasedScanJobRunner uses Celery to parallelize some of
+        the tasks that don't necessarily depend on each other. For a ScanJob that has
+        multiple Sources, the ScanTasks may run in parallel for separate Sources. After
+        *all* of them have completed, Celery tasks run to fingerprint and store the
+        final state of the ScanJob.
+        """
+        if not self.scan_job.status_start():
+            error_message = "ScanJob failed to transition to running state."
+            self.scan_job.status_fail(error_message)
+            return
+
+        # Import Celery tasks locally to avoid a potential import loop.
+        # This is an unfortunate necessity (until future refactoring) because the
+        # tasks module also imports from this jobs module.
+        from scanner.tasks import (  # pylint: disable=import-outside-toplevel
+            celery_run_task_runner,
+            finalize_scan,
+            fingerprint,
+        )
+
+        # Get and group relevant IDs and types for Celery tasks to fetch later.
+        incomplete_scan_tasks = self.scan_job.tasks.filter(
+            status__in=[ScanTask.RUNNING, ScanTask.PENDING],
+            scan_type__in=[ScanTask.SCAN_TYPE_CONNECT, ScanTask.SCAN_TYPE_INSPECT],
+        ).order_by("source_id", "sequence_number")
+        celery_signatures_by_source = defaultdict(list)
+        for scan_task in incomplete_scan_tasks:
+            source_id = scan_task.source_id
+            scan_task_id = scan_task.id
+            source_type = scan_task.source.source_type
+            scan_type = scan_task.scan_type
+            signature = celery_run_task_runner.si(scan_task_id, source_type, scan_type)
+            celery_signatures_by_source[source_id].append(signature)
+
+        # Start a chain with a group that contains n chains, one chain for each Source.
+        # Each chain contains the tasks required for that Source. Typically, this will
+        # contain one "connect" task and one "inspect" task.
+        task_chain = celery.group(
+            celery.chain(*sigs) for sigs in celery_signatures_by_source.values()
+        )
```

Comment:
This is great for mimicking current thread-based manager behavior.

In a multi-source scenario we will only need to chain connect to inspect per source and have them running in parallel for each source (but we will still need to group them together to need to chain fingerprint in the end)